core/rawdb/freezer_table.go

```diff
@@ -0,0 +1,284 @@
+	// If out item count is corrent, don't do anything
+	if t.items <= items {
+		return nil
+	}
+	// Something's out of sync, truncate the table's offset index
+	t.logger.Warn("Truncating freezer table", "items", t.items, "limit", items)
+	if err := t.offsets.Truncate(int64(items+1) * 8); err != nil {
+		return err
+	}
+	// Calculate the new expected size of the data file and truncate it
+	offset := make([]byte, 8)
+	t.offsets.ReadAt(offset, int64(items)*8)
+	expected := binary.LittleEndian.Uint64(offset)
+
+	if err := t.content.Truncate(int64(expected)); err != nil {
+		return err
+	}
+	// All data files truncated, set internal counters and return
+	t.items, t.bytes = items, expected
+	return nil
+}
+
+// Close unmaps all active memory mapped regions.
+func (t *freezerTable) Close() error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	var errs []error
+	if err := t.offsets.Close(); err != nil {
+		errs = append(errs, err)
+	}
+	t.offsets = nil
+
+	if err := t.content.Close(); err != nil {
+		errs = append(errs, err)
+	}
+	t.content = nil
+
+	if errs != nil {
+		return fmt.Errorf("%v", errs)
+	}
+	return nil
+}
+
+// Append injects a binary blob at the end of the freezer table. The item index
+// is a precautionary parameter to ensure data correctness, but the table will
+// reject already existing data.
+//
+// Note, this method will *not* flush any data to disk so be sure to explicitly
+// fsync before irreversibly deleting data from the database.
+func (t *freezerTable) Append(item uint64, blob []byte) error {
+	// Ensure the table is still accessible
+	if t.offsets == nil || t.content == nil {
+		return errClosed
+	}
+	// Ensure only the next item can be written, nothing else
+	if t.items != item {
+		panic(fmt.Sprintf("appending unexpected item: want %d, have %d", t.items, item))
+	}
+	// Encode the blob and write it into the data file
+	blob = snappy.Encode(nil, blob)
+	if _, err := t.content.Write(blob); err != nil {
+		return err
+	}
+	t.bytes += uint64(len(blob))
+
+	offset := make([]byte, 8)
+	binary.LittleEndian.PutUint64(offset, t.bytes)
+	if _, err := t.offsets.Write(offset); err != nil {
+		return err
+	}
+	t.items++
+
+	t.writeMeter.Mark(int64(len(blob) + 8)) // 8 = 1 x 8 byte offset
+	return nil
+}
+
+// Retrieve looks up the data offset of an item with the given index and retrieves
+// the raw binary blob from the data file.
+func (t *freezerTable) Retrieve(item uint64) ([]byte, error) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	// Ensure the table and the item is accessible
+	if t.offsets == nil || t.content == nil {
+		return nil, errClosed
+	}
+	if t.items <= item {
+		return nil, errOutOfBounds
+	}
+	// Item reachable, retrieve the data content boundaries
+	offset := make([]byte, 8)
+	if _, err := t.offsets.ReadAt(offset, int64(item*8)); err != nil {
+		return nil, err
+	}
+	start := binary.LittleEndian.Uint64(offset)
+
+	if _, err := t.offsets.ReadAt(offset, int64((item+1)*8)); err != nil {
```

Comment:
I think this means that you can't read back the last item, since trying to read the `item +1` index will fail

Reply:
nevermind, my bad